[BB-9488] feat: prepare sumac branch

Cup0fCoffee opened the pull request:
This PR ports all code drift from `opencraft-release/redwood.1` to `opencraft-release/sumac.1`. You can see all commits that are present in `opencraft-release/redwood.1` and are missing in `open-release/redwood.master` [here](https://github.com/openedx/edx-platform/compare/open-release/redwood.master...open-craft:edx-platform:opencraft-release/redwood.1) or locally via `git log <opencraft remote>/opencraft-release/redwood.1 ^<openedx remote>/open-release/redwood.master`.

[Here](https://docs.google.com/spreadsheets/d/1iLr2GEPC2HlV_ls1UjU7hBxCFZa41-5HmpoSm2DfzDs/edit?gid=0#gid=0) you can find a table with the list of all commits and useful information that was used to determine which commits should've been cherry picked.

For reference, here is the [code drift board](https://github.com/orgs/open-craft/projects/1/views/2?filterQuery=repo%3A%22open-craft%2Fedx-platform%22%2C%22openedx%2Fedx-platform%22).

kaustavb12 committed:
feat: allow switching anonymous user ID hashing algorithm from shake to md5

The hashing algorithm has been changed in cd60646. However, there are Open edX
operators who maintain backward compatibility of anonymous user IDs after past
rotations of their Django secret key. For them, altering the hashing algorithm
was a breaking change that made their analytics inconsistent.

(cherry picked from commit 746e4fed1bfeb731b2f6288fe3ba4ea056397fdb)
(cherry picked from commit ff6d92fd892dff5e66bbf77eadf9b91ed2189a24)
(cherry picked from commit 7245bdc7f73d9693f15acc885bee4e0d00be8245)
(cherry picked from commit 6da7f588ddf0c851f49ab2f3829417fb7aec1718)
(cherry picked from commit ee3aa4acee5d05e1c05958f43b97fc1f4d59ef77)

xitij2000 committed:
temp: Add configuration option to redirect to external site when TAP account is unlinked

(cherry picked from commit e83a8c8f82849644cf95534cde3fe149e4f11916)
(cherry picked from commit 0c831dc390f63ea9259668ec400df9c4bccd95c8)
(cherry picked from commit c596bf3b100bcc629602f0c371663da2c0b862ac)
(cherry picked from commit 64f5deab99011aea3a03567bb9f4816ecde77230)

0x29a committed:
feat: all eSHE role features squashed in one commit

fix: give superusers all studio permissions

(cherry picked from commit 8ef55754f4a529cc6b784298320fcdb8b415bd83)
(cherry picked from commit 8e281a9535b8317e24f8d2b3b5d840a3eb852865)
(cherry picked from commit f55297379276ddcdfd140b43bdc54d19128744d9)
(cherry picked from commit 6de7b64ae477bf882c1ae3a97ef2e8940e773a57)

feat: eSHE Instructor role

Adds the eSHE Instructor role, which inherits Course Staff permissions,
but isn't able to enroll / un-enroll students and can't assing course
team roles unless in combination with Course Staff / Instructor /
Discussion admin roles.

(cherry picked from commit 5d160c2ecd91e58276dccfd84ca32fb689f46e5e)
(cherry picked from commit a21b4f0da1d6eb5240d15aa8b50fd77454b47167)

feat: Teaching Assistant role

(cherry picked from commit 176de06291aafe4a8f8c9a3a9afdec24cedb332e)
(cherry picked from commit 7ef00c0dd265c09a52c71feab9d58a34f7f2c54c)
(cherry picked from commit a0d88065b11f8d182d3c00deb5935880bf0611e1)

Kelketek committed:
feat: Allow trusted apps to perform cookie login.

(cherry picked from commit 7fcd1988c5e51b3449c83f9047ed6c91ea576ee4)

viadanna committed:
feat: DEFAULT_COURSE_INVITATION_ONLY allow for invitation_only new courses by default

(cherry picked from commit ffe3ff4ba098c748c8fb5f56042cd4873d6b024b)
(cherry picked from commit bb9c2bcc0e2452a9a11fe81ae812c81e8c22cdc2)

Cup0fCoffee committed:
feat: add a feature flag to disable dates tab for all courses

(cherry picked from commit 88bb2a4cabd4c5d49efcb9e44097a765e1fe433d)

CefBoud committed:
feat: tpa automatic logout with a single redirect

(cherry picked from commit e8c942c77d46a5e5c4b7afc5904c130897cdffe2)
(cherry picked from commit 622d891a5b036f416c270e1a1665991f1357cc87)

tecoholic committed:
fix: prevent redirects to /undefined after saml auth

When a user authenticates using SAML, but their accounts aren't linked,
they land on the login page with the message to link their accounts. If the
`OC_REDIRECT_ON_TPA_UNLINKED_ACCOUNT` value is not set, they are redirected to
`/undefined`. This commit checks that the `redirectURL` is a valid non-empty
string before performing the redirect.

The bug was introduced in 91f8d36, so this can be safely dropped once that
temporary commit is removed from our branches. For Redwood, the relevant
commit hash is 64f5deab99.

Internal-ref: https://tasks.opencraft.com/browse/BB-9010
(cherry picked from commit 9ade4eac447f5b90d271b26cfee1c852c2e3c559)
(cherry picked from commit b801f99876ae53abb6fcf0d2dc04c93a76182a66)

kaustavb12 committed:
feat: in-context discussion for units can be disabled by default (#676)

(cherry picked from commit 10aec43a69c01a3ce5e952d3c20f19141ec568a4)

viadanna committed:
fix: unhide discussion tab when enabling it (#677)

(cherry picked from commit 11ddc19e69adb8b41ede30175fa268660757b588)

viadanna committed:
fix: unhide discussion tab when enabling it (#677)

(cherry picked from commit 80e27f6d2e915c033344d4d2d9e14614876a8769)

Agrendalath committed:
fix: remove incorrect key type when filtering inaccessible blocks

Returning the list with a 'completion' string was added in 9bc0f85. However,
the `get_accessible_sequences` method expects a list of dicts with an 'id' key.

(cherry picked from commit be897f0a9a073fb35c5722172bdf557efd5d8ac1)

0x29a committed:
feat: add has_course_author_access to CourseHomeMetadataView response (#682)

Cherry-picked from https://github.com/openedx/edx-platform/pull/35313

(cherry picked from commit 50470c1e5804899742899cef49b70cddabdcb26a)

0x29a committed:
feat: allow Bearer auth for sequence metadata

(cherry picked from commit 3ba4ab1a46820c5fe6ccf1c4592a780f04c6c1b6)

feanil committed:
Merge pull request #35713 from openedx/feanil/ubuntu-24.04

feanil/ubuntu 24.04

(cherry picked from commit 642dae67f1be5a6e49adbd04ea51f2857bf9d3d3)

xitij2000 committed:
feat: User agreements API for generic agreement records

This change adds a new kind of generic user agreement that allows plugins or
even the core platform to record a user's acknowledgement of an agreement.

(cherry picked from commit dc0383f6e38a3a7eb8e82ea1b18d0baaf66192c1)
(cherry picked from commit 2481c0e670b58df2e74a6725c3a0a6ac7bd4ac32)

0x29a committed:
fix: use a single 'provider_type' key for storing discussion provider type in course (#722)

Both 'provider' and 'provider_type' have been used for storing the discussion provider type in course 'discussions_settings' field, there are some places in the code checking for 'provider' and others checking for 'provider_type', in some cases this can cause a bug where it doesn't detect the correct provider which causes discussion settings not being copied correctly when a course is cloned.

This change prioritises the `provider_type` setting over `provider` and reads `provider` only as a fallback. The `provider` setting is now made read-only just for backwards-compatibility, to avoid confusion.

(cherry picked from commit cbddd4ef84e90669c1d00b1416c3c0b2850bb2fa)

Co-authored-by: kshitij.sobti <kshitij@sobti.in>
(cherry picked from commit 8612f2acfaca1a312830a3bf5755b6868ec0c92a)

0x29a reviewed (approved):
:+1:

- [ ] I didn't test this. Only checked that the instance is working (registration, enrollments, etc). The reason is in the ticket comments.
- [x] I read through the list of commits in the spreadsheet and compared it to the actual diff.

viadanna committed:
chore: bump ora2 fork

Setting ora2 dependency to our fork while awaiting upstream PR being
merged[1]. Once it's merged, upstream will bump up the dependency
version which will include this change, so will be able to drop this
commit.

1: https://github.com/openedx/edx-ora2/pull/2237

(cherry picked from commit 506201e12f703d08b9c8c71b7d08d6deb0b43ce7)
